go/libraries/doltcore/diff/querydiff/analyze.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/liquidata-inc/go-mysql-server/sql"
+	"github.com/liquidata-inc/go-mysql-server/sql/plan"
+)
+
+// lazyQueryPlan transforms a query plan by removing Project nodes and returning
+// the composite projections of the plan tree. As the plan is transformed, Expressions that
+// rely on these projections are transformed to lazily evaluate the projections.
+func lazyQueryPlan(node sql.Node) (lazyNode sql.Node, projections []sql.Expression, err error) {
+	// ignore opaque nodes
+
+	if tbl, ok := node.(sql.Table); ok {
+		return node, getFieldsForTable(tbl), nil
+	}
+
+	children := node.Children()
+	if len(children) == 0 {
+		return nil, nil, fmt.Errorf("reached bottom of query plan unexpectedly")
+	}
+
+	offset := 0
+	lazyChildren := make([]sql.Node, len(children))
+	for i, c := range children {
+		c, pjs, err := lazyQueryPlan(c)
+		if err != nil {
+			return nil, nil, err
+		}
+		lazyChildren[i] = c
+
+		pjs = shiftFieldIndexes(pjs, offset)
+		projections = append(projections, pjs...)
+		offset += len(c.Schema())
+	}
+
+	node, err = node.WithChildren(lazyChildren...)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	lazyNode, err = plan.TransformExpressions(node, func(e sql.Expression) (sql.Expression, error) {
+		return makeExpressionLazy(e, projections)
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if p, ok := lazyNode.(*plan.Project); ok {
+		lazyNode = p.Child
+		projections = p.Expressions()
+	}
+
+	return lazyNode, projections, nil
+}
+
+func extractRowOrder(node sql.Node) (composite []plan.SortField) {
+	if t, ok := node.(sql.Table); ok {
+		for i, col := range t.Schema() {
+			if !col.PrimaryKey {
+				continue
+			}
+			pkOrder := plan.SortField{
+				Column:       expression.NewGetField(i, col.Type, col.Name, col.PrimaryKey),
+				Order:        plan.Ascending,
+				NullOrdering: plan.NullsFirst,
+			}
+			composite = append(composite, pkOrder)
+		}
+		return composite
+	}
+
+	for _, c := range node.Children() {
+		sf := extractRowOrder(c)
+		composite = append(composite, sf...)
+	}
+
+	if s, ok := node.(*plan.Sort); ok {
+		// todo: prune sort fields
+		// having columns duplicated here won't lead to incorrect
+		// results, but could lead to extra work during sorting
+		composite = append(s.SortFields, composite...)
+	}
+
+	return composite
+}
+
+func makeExpressionLazy(e sql.Expression, composite []sql.Expression) (sql.Expression, error) {
+	gf, ok := e.(*expression.GetField)
+	if ok {
+		if gf.Index() >= len(composite) {
+			return nil, fmt.Errorf("index out of bounds in lazy expression substitution")
+		}
+		e = composite[gf.Index()]
+	}
+	return e, nil
+}
+
+func getFieldsForTable(tbl sql.Table) []sql.Expression {
+	fields := make([]sql.Expression, len(tbl.Schema()))
+	for i, col := range tbl.Schema() {
+		fields[i] = expression.NewGetField(i, col.Type, col.Name, col.PrimaryKey)
+	}
+	return fields
+}
+
+func shiftFieldIndexes(composite []sql.Expression, offset int) []sql.Expression {
```

Comment:
Check out fix_field_indexes.go to see if there is anything you can use there by making it publicly exported. I think it's a more general solution to what you're trying to do here with offsets.